arangod/Aql/SortExecutor.cpp

```diff
@@ -179,3 +228,59 @@ std::pair<ExecutionState, size_t> SortExecutor::expectedNumberOfRows(size_t atMo
   }
   return {ExecutionState::DONE, rowsLeft};
 }
+
+template<typename OutputRowImpl>
+std::tuple<ExecutionState, NoStats, size_t> SortExecutor<OutputRowImpl>::skipRows(size_t toSkip) {
+  if (_input == nullptr) {
+    auto fetchRes = consumeInput();
+    if (fetchRes.first == ExecutionState::WAITING) {
+      return std::make_tuple(fetchRes.first, fetchRes.second, 0);
+    }
+  }
+  const size_t skipped = std::min(toSkip, _sortedIndexes.size() - _returnNext);
+  _returnNext += skipped;
+  return std::make_tuple(
+    _returnNext >= _sortedIndexes.size() ? ExecutionState::DONE : ExecutionState::HASMORE,
+    NoStats(), skipped);
+}
+
+
+arangodb::IndexIterator::DocumentCallback MaterializerProducer::ReadContext::copyDocumentCallback(ReadContext & ctx) {
+  auto* engine = EngineSelectorFeature::ENGINE;
+  TRI_ASSERT(engine);
+  typedef std::function<arangodb::IndexIterator::DocumentCallback(ReadContext&)> CallbackFactory;
+  static CallbackFactory const callbackFactories[]{
+    [](ReadContext& ctx) {
+    // capture only one reference to potentially avoid heap allocation
+    return [&ctx](LocalDocumentId /*id*/, VPackSlice doc) {
+      TRI_ASSERT(ctx._outputRow);
+      TRI_ASSERT(ctx._inputRow);
+      TRI_ASSERT(ctx._inputRow->isInitialized());
+      TRI_ASSERT(ctx._infos);
+      arangodb::aql::AqlValue a{ arangodb::aql::AqlValueHintCopy(doc.begin()) };
+      bool mustDestroy = true;
+      arangodb::aql::AqlValueGuard guard{ a, mustDestroy };
+      ctx._outputRow->moveValueInto(ctx._infos->outputMaterializedDocumentRegId(), *ctx._inputRow, guard);
+    };
+  },
+
+    [](ReadContext& ctx) {
```

Comment:
you could write this stuff as a standalone function and I think it would be more readable